File: csharp/Tdcv2/Output/AtomicFile.cs
using System;
using System.IO;

namespace Tdcv2.Output;

/// <summary>Writing a run's output so a failed run leaves the destination as it found it.</summary>
/// <remarks>
/// The output is written to <c>&lt;path&gt;.partial</c> beside the destination and moved over it
/// only on <see cref="Commit"/>; disposed without a commit, the partial file is removed and the
/// destination is untouched. The move is within one directory, so within one filesystem, and
/// therefore atomic: a reader sees the old file or the new one, never half of either.
/// <para>
/// Measured before this class, on all five implementations: a run that failed with
/// <c>-o out.csv</c> destroyed an existing <c>out.csv</c> — four truncated it to nothing, one
/// deleted it.
/// </para>
/// <para>
/// A symbolic link is resolved first and its TARGET written, so the link stays a link. A device
/// (<c>-o /dev/stdout</c>) cannot be replaced by a move, so it is written directly, by the name it
/// was given — and so is any destination whose directory refuses a file beside it. .NET 6 cannot
/// ask Unix for a file's type, so a device is recognised by living under <c>/dev/</c>; a named pipe
/// elsewhere would be replaced by a regular file, where the reference writes into it.
/// </para>
/// </remarks>
internal sealed class AtomicFile : IDisposable
{
    private readonly string? _temp;
    private readonly string _target;
    private bool _done;

    private AtomicFile(FileStream stream, string? temp, string target)
    {
        Stream = stream;
        _temp = temp;
        _target = target;
    }

    /// <summary>Where the output goes until it is committed.</summary>
    internal FileStream Stream { get; }

    /// <summary>The partial file beside a destination.</summary>
    internal static string PartialPath(string path) => path + ".partial";

    /// <summary>Open <paramref name="path"/> for a run's output.</summary>
    internal static AtomicFile Open(string path)
    {
        string target = Path.GetFullPath(path);
        if (File.Exists(target)
            && new FileInfo(target).ResolveLinkTarget(returnFinalTarget: true) is { } real)
        {
            target = real.FullName;
        }

        bool device = target.StartsWith("/dev/", StringComparison.Ordinal);
        if (!device)
        {
            string temp = PartialPath(target);
            try
            {
                return new AtomicFile(Create(temp), temp, target);
            }
            catch (Exception e) when (e is IOException or UnauthorizedAccessException)
            {
                // No file can be made beside it: written in place, below.
            }
        }

        return new AtomicFile(Create(path), null, target);
    }

    /// <summary>The finished output takes the destination's place.</summary>
    internal void Commit()
    {
        Stream.Dispose();
        _done = true;
        if (_temp is not null)
        {
            File.Move(_temp, _target, overwrite: true);
        }
    }

    /// <summary>Without a commit: the partial file goes, and the destination stays as it was.</summary>
    public void Dispose()
    {
        if (_done)
        {
            return;
        }

        _done = true;
        Stream.Dispose();
        if (_temp is not null)
        {
            File.Delete(_temp);
        }
    }

    private static FileStream Create(string path) =>
        new(path, FileMode.Create, FileAccess.Write, FileShare.None, 1 << 16);
}

File: java/src/main/java/io/github/nickliapin/tdc/output/AtomicFile.java
package io.github.nickliapin.tdc.output;

import java.io.Closeable;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;

/**
 * Writing a run's output so a failed run leaves the destination as it found it.
 *
 * <p>The output is written to {@code <path>.partial} beside the destination and moved over it only
 * on {@link #commit()}; closed without a commit, the partial file is removed and the destination
 * is untouched. The move is within one directory, so within one filesystem, and therefore atomic: a
 * reader sees the old file or the new one, never half of either.
 *
 * <p>Measured before this class, on all five implementations: a run that failed with {@code -o
 * out.csv} destroyed an existing {@code out.csv} — four truncated it to nothing, one deleted it.
 *
 * <p>A symbolic link is resolved first and its TARGET written, so the link stays a link. A
 * destination that exists and is not a regular file ({@code -o /dev/stdout}, a named pipe) cannot
 * be replaced by a move, so it is written directly, by the name it was given — and so is one whose
 * directory refuses a file beside it ({@code /dev/stdout} redirected to a file resolves to {@code
 * /dev/fd/1}). Both are written exactly as they were before this class.
 */
public final class AtomicFile implements Closeable {
  private final OutputStream out;
  private final Path temp;
  private final Path target;
  private boolean done;

  private AtomicFile(OutputStream out, Path temp, Path target) {
    this.out = out;
    this.temp = temp;
    this.target = target;
  }

  /** The partial file beside a destination. */
  public static Path partialPath(Path path) {
    return path.resolveSibling(path.getFileName() + ".partial");
  }

  /** Open {@code path} for a run's output. */
  public static AtomicFile open(Path path) throws IOException {
    boolean exists = Files.exists(path);
    Path target = exists ? path.toRealPath() : path;
    if (!exists || Files.isRegularFile(target)) {
      Path temp = partialPath(target);
      try {
        return new AtomicFile(Files.newOutputStream(temp), temp, target);
      } catch (IOException e) {
        // No file can be made beside it: written in place, below.
      }
    }
    return new AtomicFile(Files.newOutputStream(path), null, target);
  }

  /** Where the output goes until it is committed. */
  public OutputStream stream() {
    return out;
  }

  /** The finished output takes the destination's place. */
  public void commit() throws IOException {
    out.close();
    done = true;
    if (temp == null) {
      return;
    }
    try {
      Files.move(
          temp, target, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
    } catch (AtomicMoveNotSupportedException e) {
      Files.move(temp, target, StandardCopyOption.REPLACE_EXISTING);
    }
  }

  /** Without a commit: the partial file goes, and the destination stays as it was. */
  @Override
  public void close() throws IOException {
    if (done) {
      return;
    }
    done = true;
    try {
      out.close();
    } finally {
      if (temp != null) {
        Files.deleteIfExists(temp);
      }
    }
  }
}
